grove/tomography/utils.py

```diff
@@ -0,0 +1,640 @@
+
+EPS = 1e-8
+
+
+def sample_outcomes(probs, n):
+    """
+    For a discrete probability distribution ``probs`` with outcomes 0, 1, ..., k-1 draw ``n``
+    random samples.
+
+    :param list probs: A list of probabilities.
+    :param Number n: The number of random samples to draw.
+    :return: An array of samples drawn from distribution probs over 0, ..., len(probs) - 1
+    :rtype: numpy.ndarray
+    """
+    dist = np.cumsum(probs)
+    rs = np.random.rand(n)
+    return np.array([(np.where(r < dist)[0][0]) for r in rs])
+
+
+def basis_state_preps(*qubits):
+    """
+    Generate a sequence of programs that prepares the measurement
+    basis states of some set of qubits in the order such that the qubit
+    with highest index is iterated over the most quickly:
+    E.g., for qubits=(0, 1), it returns the circuits::
+
+        I_0 I_1
+        I_0 X_1
+        X_0 I_1
+        X_0 X_1
+    :param list qubits: Each qubit to include in the basis state preparation.
+    :return: Yields programs for each basis state preparation.
+    :rtype: Program
+    """
+    for prep in cartesian_product([I, X], repeat=len(qubits)):
+        basis_prep = Program()
+        for gate, qubit in zip(prep, qubits):
+            basis_prep.inst(gate(qubit))
+        yield basis_prep
+
+
+def basis_labels(n):
+    """
+    Generate a list of basis labels for `n` qubits, ordered from least to greatest, in big-endian
+     format:
+
+        ['00..00', '00..01', ..., '11..11']
+
+    :param n:
+    :return: A list of strings of length n that enumerate the n-qubit bitstrings
+    :rtype: list
+    """
+    return ["".join(labels) for labels in itertools.product('01', repeat=n)]
+
+
+def sample_bad_readout(program, num_samples, assignment_probs, cxn):
+    """
+    Generate `n` samples of measuring all outcomes of a Quil `program`
+    assuming the assignment probabilities `assignment_probs` by simulating the
+    wave function on a reference-qvm SyncConnection `cxn`
+
+    :param pyquil.quil.Program program: The program.
+    :param int num_samples: The number of samples
+    :param numpy.ndarray assignment_probs: A matrix of assignment probabilities
+    :param reference_qvm.api.SyncConnection cxn: the QVM connection.
+    :return: The resulting sampled outcomes from assignment_probs applied to cxn, one dimensional.
+    :rtype: numpy.ndarray
+    """
+    wf = cxn.wavefunction(program)
+    return sample_outcomes(assignment_probs.dot(abs(wf.amplitudes.ravel())**2), num_samples)
+
+
+def make_histogram(samples, ksup):
+    """
+    For a list of samples [s1, s2, ..., sN] taking on integer values from 0 to ksup-1,
+    make a histogram of each integer's outcome and return it.
+
+    :param samples: The samples.
+    :param ksup: The (exclusive) upper bound
+    :return: A histogram of outcomes.
+    :rtype: numpy.ndarray
+    """
+    return np.histogram(samples, np.arange(ksup + 1) - 0.5)[0]
+
+
+def estimate_assignment_probs(bitstring_prep_histograms):
```

Comment:
if this is used to normalize the output of `make_histogram` we can just provide the `normed` kwarg to the `np.histogram` https://docs.scipy.org/doc/numpy-1.13.0/reference/generated/numpy.histogram.html

Reply:
So currently we both return the unnormalized counts, and use the normalized counts. I think having the counts is useful, so I think it would be good to keep them, and thus keep this function for normalizing it.